File: 2025/day01/day01.py
import unittest


def parse_input(filename: str) -> list[tuple[str, int]]:
    data = []

    with open(filename) as f:
        data = [line.strip() for line in f.readlines()]
    
    rotations = [
        ( line[0], int(line[1:]) )
        for line in data
    ]
    return rotations


def part1(filename: str) -> int:
    return count_zeroes(filename)


def part2(filename: str) -> int:
    return count_zeroes(filename, in_between=True)


def count_zeroes(filename: str, in_between: bool = False):
    dial: int = 50
    zeroes_at_end: int = 0
    zeroes_in_between: int = 0

    rotations = parse_input(filename)

    for (direction, value) in rotations:
        zeroes_in_between += value // 100 # full spins
        remainder = value % 100

        if direction == 'L':
            if dial != 0 and dial - remainder < 0:
                zeroes_in_between += 1
            dial = (dial - remainder) % 100
        elif direction == 'R':
            if dial + remainder > 100:
                zeroes_in_between += 1
            dial = (dial + remainder) % 100

        if dial == 0:
            zeroes_at_end += 1
    
    if in_between:
        return zeroes_at_end + zeroes_in_between
    else:
        return zeroes_at_end


class TestDay01(unittest.TestCase):
    def test_part1_example(self):
        self.assertEqual(part1("example.txt"), 3)
    
    def test_part1_input(self):
        self.assertEqual(part1("input.txt"), 964)
    
    def test_part2_example(self):
        self.assertEqual(part2("example.txt"), 6)
    
    def test_part2_input(self):
        self.assertEqual(part2("input.txt"), 5872)


if __name__ == "__main__":
    print("Part 1:", part1("input.txt"))
    print("Part 2:", part2("input.txt"))

    unittest.main()

File: 2025/day01/day01.nim
import std/[strutils, math]


type
    Direction = enum
        Left = "L"
        Right = "R"

    Rotation = object
        direction: Direction
        value: int


proc parse_data(filename: string): seq[Rotation] =
    for line in filename.readFile.splitLines:
        result.add Rotation(
            direction: parseEnum[Direction]($line[0]),
            value: parseInt(line[1..line.high])
        )


proc count_zeroes(filename: string, count_between: bool = false): int =
    let rotations: seq[Rotation] = parse_data(filename)
    var
        dial: int = 50
        remainder: int
        in_between: int
    
    for rotation in rotations:
        in_between += rotation.value.euclDiv 100
        remainder = rotation.value.euclMod 100

        case rotation.direction
        of Left:
            if dial != 0 and dial - remainder < 0:
                inc in_between
            dial = (dial - remainder).euclMod 100
        of Right:
            if dial + remainder > 100:
                inc in_between
            dial = (dial + remainder).euclMod 100
        
        if dial == 0:
            inc result
    
    if count_between:
        result += in_between


proc part1(filename: string): int =
    result = count_zeroes(filename)
    echo "Part 1 (", filename, "): ", result


proc part2(filename: string): int =
    result = count_zeroes(filename, count_between = true)
    echo "Part 2 (", filename, "): ", result


when isMainModule:
    doAssert part1("example.txt") == 3
    doAssert part1("input.txt") == 964

    doAssert part2("example.txt") == 6
    doAssert part2("input.txt") == 5872
